spec/notification_center_spec.rb

```diff
@@ -0,0 +1,484 @@
+        
+        @inner_notification_center.notifications.each do |key, _array|
+          expect(@inner_notification_center.notifications[key]).to be_empty
+        end
+      end
+
+      it 'clean_all_notifications does not throw an error when called again' do
+        @inner_notification_center.clean_all_notifications
+        expect { @inner_notification_center.clean_all_notifications }
+         .to_not raise_error
+      end
+    end
+    
+    describe '.fire_notifications' do
+      let(:raise_error_handler) { Optimizely::RaiseErrorHandler.new }
+      let(:notification_center) { Optimizely::NotificationCenter.new(spy_logger, raise_error_handler) }
+      before(:example) do
+        config = Optimizely::ProjectConfig.new(config_body_JSON, spy_logger, error_handler)
+        @event_builder = Optimizely::EventBuilder.new(config)
+        @args = [
+          config.get_experiment_from_key('test_experiment'),
+          'test_user',
+          {},
+          '111128',
+          @event_builder.create_impression_event(
+            config.get_experiment_from_key('test_experiment'),
+            '111128', 'test_user', nil
+          )
+        ]
+      end
+
+      it 'should return success log for notification sent' do
+        notification_type = Optimizely::NotificationCenter::NOTIFICATION_TYPES[:ACTIVATE]
+        notification_center.add_notification_listener(notification_type, @callback_reference)
+        notification_center.notifications[notification_type].each do |notification|
+          notification_callback = notification[:callback]
+          expect {notification_callback.call(@args) }.to_not raise_error
+        end
+        expect(notification_center.fire_notifications(notification_type, @args))
+        expect(spy_logger).to have_received(:log).once
+                               .with(Logger::INFO, "Notification #{notification_type} sent successfully.")
+      end
+
+      it 'should return nil when notification type not valid' do
+        notification_type = Optimizely::NotificationCenter::NOTIFICATION_TYPES[:ACTIVATE]
+        notification_center.add_notification_listener(notification_type, @callback_reference)
+        expect { notification_center.fire_notifications('test_type', @args) }
+         .to raise_error(Optimizely::InvalidNotificationType)
+      end
+
+      it 'should return nil and log when args are invalid' do
+        notification_type = Optimizely::NotificationCenter::NOTIFICATION_TYPES[:ACTIVATE]
+        notification_center.add_notification_listener(notification_type, @callback_reference)
+        expect(notification_center.fire_notifications(notification_type)).to eq(nil)
+        expect(spy_logger).to have_received(:log).once
+          .with(
+            Logger::ERROR,
+            'Problem calling notify callback. Error: wrong number of arguments (given 0, expected 1)'
+          )
+      end
+
+      it 'should fire multiple notifications for a single type' do
```

Comment:
In all of the fireNotifications, assert that the callback is executed by spying the callback object rather than asserting log messages